### PR TITLE
feat: add budget_limits (multi-window budget) support for users

### PR DESCRIPTION
**Pull request by rophy**
## TLDR

Problem this solves:

- Users can only have a single budget window (max_budget + budget_duration), unlike keys and teams which support multiple concurrent windows via budget_limits
- No way to enforce e.g. a $5/day AND $100/month cap on the same user simultaneously

How it solves it:

- Adds budget_limits field to the User entity, reusing the same BudgetLimitEntry model already used by keys and teams
- Hooks into existing budget enforcement (auth_checks), spend tracking (budget_reservation), and reset (reset_budget_job) paths
- Respects skip_user_budget_on_team_key setting, matching legacy single-budget behavior

## Pre-Submission checklist

- [x] I have added meaningful tests
- [x] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [x] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Verified at commit e120d21ea7 against a live proxy (postgres + redis, gemini-2.5-flash hitting real Gemini API, `PROXY_BUDGET_RESCHEDULER_MIN_TIME=10 PROXY_BUDGET_RESCHEDULER_MAX_TIME=15`)

A single user is created with a 30s/$0.000003 window and a 1h/$0.000006 window. One gemini call costs ~$0.000003, so the 30s window fills after 1 call and the 1h window fills after 2 calls

```bash
curl -s http://localhost:4000/user/new \
  -H "Authorization: Bearer sk-1234" \
  -H "Content-Type: application/json" \
  -d '{
    "budget_limits": [
      {"budget_duration": "30s", "max_budget": 0.000003},
      {"budget_duration": "1h", "max_budget": 0.000006}
    ]
  }'

curl -s http://localhost:4000/key/generate \
  -H "Authorization: Bearer sk-1234" \
  -H "Content-Type: application/json" \
  -d '{"user_id": "<USER_ID>"}'
```

| Step | Action | Expected | Actual |
|------|--------|----------|--------|
| 1 | Send one chat completion | 200 OK | OK (gemini calls: 1) |
| 2 | Send another call immediately | blocked by **30s** window | BLOCKED by 30s window |
| 3 | Wait for 30s window reset (~70s with scheduler lag) | 200 OK | reset after 70s - OK (gemini calls: 2) |
| 4 | Send another call immediately | blocked by **30s** window | BLOCKED by 30s window |
| 5 | Wait for 30s window reset again | blocked by **1h** window (30s cleared but 1h still full) | BLOCKED by 1h window (not 30s) after 65s |

Steps 2 and 4 prove the 30s window enforces. Step 3 proves the 30s window resets. Step 5 proves the 1h window enforces independently; even though the 30s window had reset, the 1h window accumulated spend from both successful calls and blocked the request. Only 2 Gemini API calls were consumed (blocked calls are rejected by the proxy before reaching the provider)

Raw output:

```
=== Setup: user with 30s/$0.000003 + 1h/$0.000006 ===
  user=af4d36ef-1550-4281-9626-a28a79fa1baf

=== Step 1: first call - should succeed ===
  OK (gemini calls: 1)

=== Step 2: call again - should be blocked by 30s ===
  BLOCKED by 30s window

=== Step 3: wait for 30s reset, call should succeed ===
  waiting up to 90s for 30s window reset...
  reset after 70s - OK (gemini calls: 2)

=== Step 4: call again - should be blocked by 30s ===
  BLOCKED by 30s window

=== Step 5: wait for 30s reset - should be blocked by 1h ===
  waiting up to 90s for 30s window reset...
  BLOCKED by 1h window (not 30s) after 65s

=== Result: 2 gemini calls used (expected: 2) ===
```

## Type

New Feature

## Changes

The change mirrors how budget_limits already works for keys and teams, extending it to users

`_user_multi_budget_check` in auth_checks.py iterates the user's budget windows, reads each window's spend counter from the in-memory cache, and returns a 400 if any window is exceeded; it respects `skip_user_budget_on_team_key`, matching legacy single-budget behavior

After a successful completion, budget_reservation.py adds the spend delta to each of the user's per-window counters (keyed `user:{id}:budget:{duration}`) alongside the existing key and team window counters; also respects `skip_user_budget_on_team_key`

The existing `_reset_entity_budget_windows` in reset_budget_job.py (which already handles keys and teams) is called for users too, zeroing expired window counters and advancing `reset_at`

spend_counter_reseed.py includes users with budget_limits in the reseed query so counters survive proxy restarts

`budget_limits` is accepted on `POST /user/new` and `POST /user/update` in internal_user_endpoints.py; proxy_server.py passes it through on the update path

All three schema.prisma files get the `budget_limits Json?` column on `LiteLLM_UserTable` plus a migration SQL; dashboard types are regenerated

Unit tests cover enforcement logic (including team key skip behavior), reset job, spend counter reseed, and management endpoint parsing; an e2e test creates a user with a 30s window, drives calls until blocked, waits for the reset job, and verifies the window resets

## QA runbook

- tests/e2e/quota_management/budgets/test_user_multi_window_budget_e2e.py::test_user_short_window_blocks_then_resets - a user with a 30s/$0.000003 window gets blocked after a few calls, then unblocked after the window resets
  - [ ] Create a user with budget_limits containing a 30s window capped at $0.000003 and a 1m window capped at $1.00
  - [ ] Generate a personal key (no team_id) for that user
  - [ ] Send chat completions (claude-haiku-4-5, max_tokens=16) in a loop with 2s spacing until a 400 budget error is returned
  - [ ] Verify the block happened (the 30s window was enforced)
  - [ ] Continue polling every 5s until a 200 is returned (the reset job cleared the expired window)
  - [ ] Verify the reset happened within 90s of the window start (30s window + scheduler lag)
  - [ ] Sanity check: this test makes sense to add and is not hand-wavey (e.g., assert actual expected spend instead of just spend > 0) or potentially flaky

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

### CODE REVIEW

**Comment by rophy** [comment]
@greptileai

File: litellm/proxy/proxy_server.py [hunk 1]
```diff
@@ -2450,13 +2450,32 @@ async def _team_member_scope(scope_user_id: str, scope_team_id: str) -> None:
...
+                for window in user_budget_limits:
+                    duration = window["budget_duration"] if isinstance(window, dict) else window.budget_duration
+                    user_window_counter = f"spend:user:{scope_user_id}:window:{duration}"
+                    if user_window_counter not in reserved_counter_keys:
+                        await _init_and_increment_window_spend_counter(
+                            counter_key=user_window_counter,
+                            entity_type="User",
+                            entity_id=scope_user_id,
+                            window_start=get_budget_window_start(window),
+                            increment=cost,
```
[the remark this answers is not in the document]
Comment (rophy) [reply litellm/proxy/proxy_server.py]: This is consistent with how the legacy single-budget path works. The \`spend:user:{id}\` counter at L2452-2458 in the same function is also incremented unconditionally for team-key requests; \`skip_user_budget_on_team_key\` only controls enforcement (the budget comparison in \`_user_max_budget_check\`), not spend tracking. The multi-window counters follow the same pattern

Changing this to conditionally skip counter increments would alter existing behavior across all budget types and should be a separate PR if desired

**Comment by rophy** [comment]
@greptileai

File: litellm/proxy/management_endpoints/internal_user_endpoints.py [hunk 2]
```diff
@@ -1117,6 +1117,20 @@ def _update_internal_user_params(
...
 
+    if "budget_limits" in non_default_values:
+        raw_windows = non_default_values["budget_limits"]
+        if raw_windows:
+            from litellm.proxy.common_utils.timezone_utils import get_budget_reset_time
+
+            initialized_windows = []
+            for window in raw_windows:
+                w = window if isinstance(window, dict) else window.model_dump()
+                w["reset_at"] = get_budget_reset_time(budget_duration=w["budget_duration"]).isoformat()
```
[the remark this answers is not in the document]
Comment (rophy) [reply litellm/proxy/management_endpoints/internal_user_endpoints.py]: This is the same pattern used by the key update path at key_management_endpoints.py:1942-1947, which also calls `get_budget_reset_time` without clearing the spend counter. The user update code here is intentionally consistent with the existing key/team behavior

Fixing this across all entity types (keys, teams, users) would be a worthwhile follow-up but is out of scope for this PR, which adds user support on top of the existing infrastructure without changing its semantics

**Comment by rophy** [comment]
@greptileai

**Comment by rophy** [comment]
@greptileai

File: litellm/proxy/management_endpoints/internal_user_endpoints.py [hunk 3]
```diff
@@ -1120,6 +1120,20 @@ def _update_internal_user_params(
...
+        raw_windows: list[BudgetLimitEntry | dict[str, object]] | None = non_default_values["budget_limits"]
+        if raw_windows:
+            from litellm.proxy.common_utils.timezone_utils import get_budget_reset_time
+
+            initialized_windows: list[dict[str, object]] = []
+            for window in raw_windows:
+                w: dict[str, object] = window if isinstance(window, dict) else window.model_dump()
+                w["reset_at"] = get_budget_reset_time(budget_duration=str(w["budget_duration"])).isoformat()
+                initialized_windows.append(w)
+            non_default_values["budget_limits"] = json.dumps(initialized_windows)
```
[the remark this answers is not in the document]
Comment (rophy) [reply litellm/proxy/management_endpoints/internal_user_endpoints.py]: The user_update endpoint is decorated with @management_endpoint_wrapper (internal_user_endpoints.py:1384), which calls _delete_user_id_from_cache after a successful update (management_helpers/utils.py:642). That function deletes the user entry from user_api_key_cache by user_id (management_helpers/utils.py:371), so the next auth check re-fetches from DB and picks up the new budget_limits

File: litellm/proxy/common_utils/reset_budget_job.py [hunk 4]
```diff
@@ -722,51 +726,69 @@ async def reset_budget_windows(self) -> None:
...
+                where={"team_id": row_id},
+                data={"budget_limits": windows_json},  # type: ignore[arg-type]
+            ),
+            entity_label="teams",
+            spend_counter_cache=spend_counter_cache,
+            now=now,
+        )
 
-        # --- Teams ---
+        await self._reset_entity_budget_windows(
```
[the remark this answers is not in the document]
Comment (rophy) [reply litellm/proxy/common_utils/reset_budget_job.py:758]: The timezone stripping on line 701 is in _reset_expired_window, a shared function that already handled keys and teams before this PR. Our change (line 758) just adds users as another caller of the same function. The concern is valid but pre-existing and applies equally to keys/teams; fixing it across all entity types would be a separate PR
